python/pyarrow/array.pxi

```diff
@@ -161,7 +161,9 @@ def array(object obj, type=None, mask=None, size=None, from_pandas=None,
     else:
         c_from_pandas = from_pandas
 
-    if _is_array_like(obj):
+    if hasattr(obj, '__arrow_array__'):
+        return obj.__arrow_array__(type=type)
```

Comment:
Would it make sense to check for a return value of `NotImplemented` and fall back to the default path?

Reply:
I don't think so. Why would you define `__arrow_array__` just to return `NotImplemented`?

Reply:
Yes, I think if you don't want to implement this, the user should simply not define `__arrow_array__`. 

But we should probably add a check to ensure that was is returned from `__arrow_array__` is actually a pyarrow Array (that is something that numpy also does).